### PR TITLE
Implement #2638: [Architecture] Design Model Serialization and Chec

### LINKED ISSUE #2638: [Architecture] Design Model Serialization and Checkpoint System
## Problem

Need consistent architecture for model saving/loading and checkpointing across different model types.

### 1. State Dict Interface

```mojo
trait Serializable:
    """Interface for serializable objects (models, optimizers)."""
    
    fn state_dict(self) -> Dict[String, ExTensor]:
        """Get state as dictionary of tensors."""
    
    fn load_state_dict(mut self, state: Dict[String, ExTensor]):
        """Load state from dictionary."""

struct Model(Serializable):
    """Example model implementation."""
    var layer1: Linear
    var layer2: Linear
    
    fn state_dict(self) -> Dict[String, ExTensor]:
        var state = Dict[String, ExTensor]()
        state["layer1.weight"] = self.layer1.weight
        state["layer1.bias"] = self.layer1.bias
        state["layer2.weight"] = self.layer2.weight
        state["layer2.bias"] = self.layer2.bias
        return state
    
    fn load_state_dict(mut self, state: Dict[String, ExTensor]):
        self.layer1.weight = state["layer1.weight"]
        self.layer1.bias = state["layer1.bias"]
        self.layer2.weight = state["layer2.weight"]
        self.layer2.bias = state["layer2.bias"]
```

### 2. Checkpoint Manager

```mojo
struct CheckpointManager:
    """Manage model checkpoints."""
    var checkpoint_dir: String
    var keep_last_n: Int  # Keep only N most recent checkpoints
    
    fn save_checkpoint(
        self,
        epoch: Int,
        model: Serializable,
        optimizer: Serializable,
        metrics: Dict[String, Float32]
    ) raises:
        """Save checkpoint to disk.
        
        Saves:
        - Model state dict
        - Optimizer state dict  
        - Training metadata (epoch, metrics)
        """
        var checkpoint = Dict()
        checkpoint["epoch"] = epoch
        checkpoint["model_state"] = model.state_dict()
        checkpoint["optimizer_state"] = optimizer.state_dict()
        checkpoint["metrics"] = metrics
        
        var filepath = self.checkpoint_dir + "/checkpoint_epoch_" + str(epoch) + ".pt"
        save_to_file(checkpoint, filepath)
        
        self._cleanup_old_checkpoints()
    
    fn load_checkpoint(
        self,
        filepath: String,
        mut model: Serializable,
        mut optimizer: Serializable
    ) -> Int:
        """Load checkpoint from disk.
        
        Returns:
            Epoch number
        """
        var checkpoint = load_from_file(filepath)
        model.load_state_dict(checkpoint["model_state"])
        optimizer.load_state_dict(checkpoint["optimizer_state"])
        return checkpoint["epoch"]
    
    fn get_latest_checkpoint(self) -> Optional[String]:
        """Get path to most recent checkpoint."""
        # Find checkpoint with highest epoch number
    
    fn _cleanup_old_checkpoints(mut self):
        """Remove old checkpoints, keep only last N."""
        # Keep best + last N checkpoints
```

### 3. Model Export

```mojo
struct ModelExporter:
    """Export models to different formats."""
    
    fn export_onnx(model: Model, filepath: String, input_shape: List[Int]):
        """Export to ONNX format for interoperability."""
    
    fn export_torchscript(model: Model, filepath: String):
        """Export to TorchScript format."""
    
    fn export_mojo_package(model: Model, filepath: String):
        """Export to .mojopkg for distribution."""
```

### 4. Training Resume

```mojo
fn resume_training(
    checkpoint_path: String,
    model: Model,
    optimizer: Optimizer,
    dataloader: DataLoader
) -> Int:
    """Resume training from checkpoint.
    
    Returns:
        Epoch to resume from
    """
    var manager = CheckpointManager("./checkpoints")
    var start_epoch = manager.load_checkpoint(checkpoint_path, model, optimizer)
    
    print(f"Resuming from epoch {start_epoch}")
    return start_epoch + 1
```

## Benefits

- Consistent save/load API
- Easy training resume
- Automatic checkpoint management
- Model export to different formats
- Version compatibility checking

## File Format Considerations

- Binary format for efficiency
- JSON metadata for readability
- Version information for compatibility
- Compression for large models

## Related Features

- Best model tracking (save best validation accuracy)
- Checkpoint ensembling
- Model averaging (SWA, EMA)
- Distributed checkpointing (multi-GPU)

### PR DESCRIPTION
**Pull request by mvillmow**
Closes #2638

## Summary

I haven't made any changes yet - this is the beginning of our conversation. I'm ready to help you with tasks in the ML Odyssey repository. What would you like me to work on?

## Detailed Implementation Plan

I see that the notes directory exists but there's no `plan/` subdirectory. Based on the CLAUDE.md, planning is now done through GitHub issues. Let me provide the comprehensive plan directly in my response instead of writing to a file:

---

# Implementation Plan: Model Serialization and Checkpoint System

## Issue #2638: [Architecture] Design Model Serialization and Checkpoint System

---

## 1. Summary

This plan designs a comprehensive model serialization and c...

### CODE REVIEW

**Comment by mvillmow** [comment]
Closing this PR due to extensive Mojo compilation errors in checkpoint_manager.mojo. The implementation needs more work to fix ownership/borrowing issues with Dict[String, ExTensor] and string operations. Will reopen with proper fixes in a follow-up PR.

Key issues that need fixing:
- Dict[String, ExTensor] cannot be implicitly copied - need to use ownership transfer
- line.strip() returns StringSlice, not String
- Several methods need 'raises' annotation
- Python interop patterns need adjustment for Mojo
